### ISSUE TITLE
F string error, permission error, and a discussion of key passthrough

### ISSUE THREAD

**Issue by azurusnova** [robinuniverse/Keebie#21]
When following the instructions, I get hung up on running the script. 
This is the output I get when running "python keepbie.py" in y terminal. 

  File "keebie.py", line 73
    print(f"Untracked key {keycode} released.") # If you see this that means we missed a key press, bad news. (But not fatal.)

Unsure of how to go about to resolve this issue.

**Comment by parasbhanot** [comment]
Can you share the contents of your layer file which is handling  your key press ?

**Comment by Michael-E-B** [comment]
Firstly it would be helpful if you could proved the full stack trace and error message as this appears to only be one section of it (i.e. it does not include what error occurred).

Secondly the only issue I could imagine being caused by this line would be if you were running a version of python that didn't support f strings, please check your python version with `$ python -V`.

**Comment by Michael-E-B** [comment]
Further to my last I just ran `$ python2 ./keebie.py` and got :
```
 $ python2 ./keebie.py 
  File "./keebie.py", line 73
    print(f"Untracked key {keycode} released.") # If you see this that means we missed a key press, bad news. (But not fatal.)
                                             ^
SyntaxError: invalid syntax
```

This supports my hypothesis. Make sure you have python 3.x installed, also `keebie.py` has a shebang so you don't actually need to specify an interpreter, you should just run `./keebie.py`.

**Comment by azurusnova** [comment]
> Further to my last I just ran `$ python2 ./keebie.py` and got :
> 
> ```
>  $ python2 ./keebie.py 
>   File "./keebie.py", line 73
>     print(f"Untracked key {keycode} released.") # If you see this that means we missed a key press, bad news. (But not fatal.)
>                                              ^
> SyntaxError: invalid syntax
> ```
> 
> This supports my hypothesis. Make sure you have python 3.x installed, also `keebie.py` has a shebang so you don't actually need to specify an interpreter, you should just run `./keebie.py`.

When running it in python 3.8, this is the return I get. 

$ python3.8 keebie.py 
Traceback (most recent call last):
  File "keebie.py", line 4, in <module>
    from evdev import InputDevice, categorize, ecodes
ModuleNotFoundError: No module named 'evdev'

**Comment by azurusnova** [comment]
> Further to my last I just ran `$ python2 ./keebie.py` and got :
> 
> ```
>  $ python2 ./keebie.py 
>   File "./keebie.py", line 73
>     print(f"Untracked key {keycode} released.") # If you see this that means we missed a key press, bad news. (But not fatal.)
>                                              ^
> SyntaxError: invalid syntax
> ```
> 
> This supports my hypothesis. Make sure you have python 3.x installed, also `keebie.py` has a shebang so you don't actually need to specify an interpreter, you should just run `./keebie.py`.

After searching for evdev in my repositories (something that should have been stated that was also needed, and also running it as "sudo python3.8 keebie.py" I have gotten it to work finally. 

Running it as ./keebie.py only works when using sudo affixed to it.

**Comment by Michael-E-B** [comment]
Firstly yes, the `evdev` thing really should be documented. Second what error do you get if you run without `sudo`? If it has to do not having permission to your keyboard check out the [how to set up](https://github.com/robinuniverse/Keebie#how-to-set-up) section of the readme which has instructions for how to get permission to your board.

**Comment by azurusnova** [comment]
> Firstly yes, the `evdev` thing really should be documented. Second what error do you get if you run without `sudo`? If it has to do not having permission to your keyboard check out the [how to set up](https://github.com/robinuniverse/Keebie#how-to-set-up) section of the readme which has instructions for how to get permission to your board.

Here is the output I get. 

```
$ ./keebie.py
Welcome to Keebie
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/evdev/device.py", line 125, in __init__
    fd = os.open(dev, os.O_RDWR | os.O_NONBLOCK)
PermissionError: [Errno 13] Permission denied: '/dev/input/by-id/usb-SEMICO_USB_Keyboard-event-kbd'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "./keebie.py", line 731, in <module>
    device = InputDevice(config()[0]) # Get a reference to the keyboard on the first line of our config file
  File "/usr/lib/python3/dist-packages/evdev/device.py", line 127, in __init__
    fd = os.open(dev, os.O_RDONLY | os.O_NONBLOCK)
PermissionError: [Errno 13] Permission denied: '/dev/input/by-id/usb-SEMICO_USB_Keyboard-event-kbd'
```


And while I have figured out how to setup keys with certain outputs, like for the A key, I have it setup as CTRL+A, but when going into a program like OBS and setting up something to activate like a scene switcher and set it to the A key on the secondary keyboard, nothing comes up. Not even the A or even the binded CTRL+A that was bound to that key initially.

**Comment by Michael-E-B** [comment]
Yep, that is the sort of permission denied error that can be fixed with the steps in [how to set up](https://github.com/robinuniverse/Keebie#how-to-set-up).

With regards to your other issue that is intended behavior. Keebie grabs all inputs from the target keyboard, it does not pass through unused inputs.

**Comment by azurusnova** [comment]
> Yep, that is the sort of permission denied error that can be fixed with the steps in [how to set up](https://github.com/robinuniverse/Keebie#how-to-set-up).
> 
> With regards to your other issue that is intended behavior. Keebie grabs all inputs from the target keyboard, it does not pass through unused inputs.

So using Keebie for hotkey use on programs like OBS means that it will not work as intended?
I am to assume that Keebie will only work with commands bound to the keys, but it cannot even pass through a multiple key setup even if it is duplicated in the hotkey section of OBS?

I could also be foolish in my thinking and after setting a macro to a key, that I would need to restart Keebie so the changes take effect. Am I wrong?

**Comment by Michael-E-B** [comment]
You don't need to restart `keebie.py` for changes take effect if you edit the layer file directly, but if you want to use the TUI of `--add` or `--edit` you will need to stop `keebie.py` first because two instances can't both grab the board at the same time.

Keebie is meant to just run shell commands, scripts, and executables however if you want to have a key simulate other key(s) you can install `xdotool` and invoke it to simulate keypresses, mouse movements, and all sort of stuff. For instance I have `KEY_LEFTMETA": "xdotool key Super_L+space"` in my `default.json` layer to trigger a shortcut for my window manager.

**Comment by azurusnova** [comment]
> You don't need to restart `keebie.py` for changes take effect if you edit the layer file directly, but if you want to use the TUI of `--add` or `--edit` you will need to stop `keebie.py` first.
> 
> Keebie is meant to just run shell commands, scripts, and executables however if you want to have a key simulate other key(s) you can install `xdotool` and invoke it to simulate keypresses, mouse movements, and all sort of stuff. For instance I have `KEY_LEFTMETA": "xdotool key Super_L+space"` in my `default.json` layer to trigger a shortcut for my window manager.

https://imgur.com/a/k5LC4LK
Above I have linked a screenshot of the gui in OBS for using to bind hotkeys to different functions. 

As far as xdotool is concerned, from what I can see is that it is seriously out of my knowledgebase. I am not a coder in any means, and it looks like an utter nightmare to work with for someone inexperienced in this area like me.

**Comment by Michael-E-B** [comment]
Xdotool is pretty simple as long as you use it for simple things, for most shortcut triggering things it's as simple as `xdotool key KEYCODE_1+KEYCODE_2+...` and you can find a list of keycodes [here](https://gitlab.com/cunidev/gestures/-/wikis/xdotool-list-of-key-codes).

**Comment by azurusnova** [comment]
> Xdotool is pretty simple as long as you use it for simple things, for most shortcut triggering things it's as simple as `xdotool key KEYCODE_1+KEYCODE_2+...` and you can find a list of keycodes [here](https://gitlab.com/cunidev/gestures/-/wikis/xdotool-list-of-key-codes).

So I could use xdotool to make up bindings that I could have recognised in the hotkey section of OBS, and then bind those to keys in Keebie, and then in theory OBS would see those bindings being used when Keebie invokes xdotool to the key combinations? 

Very rube goldberg in the way of going about it, but if the keybindings in xdotool are not complicated, say like just having a combo of keys and linking them to Keebie for their respective keys, that sounds like it would work in my case. 

My whole reason for going into all of this is this was the only alternative that I have found in place of this method, problem is they no longer sell the device used in here, and the other methods do not transport well from pre-production machines to full production machines that are totally separate. (two identical servers being used for OBS video visual backbone machines)

**Comment by azurusnova** [comment]
> > Xdotool is pretty simple as long as you use it for simple things, for most shortcut triggering things it's as simple as `xdotool key KEYCODE_1+KEYCODE_2+...` and you can find a list of keycodes [here](https://gitlab.com/cunidev/gestures/-/wikis/xdotool-list-of-key-codes).
> 
> So I could use xdotool to make up bindings that I could have recognised in the hotkey section of OBS, and then bind those to keys in Keebie, and then in theory OBS would see those bindings being used when Keebie invokes xdotool to the key combinations?
> 
> Very rube goldberg in the way of going about it, but if the keybindings in xdotool are not complicated, say like just having a combo of keys and linking them to Keebie for their respective keys, that sounds like it would work in my case.
> 
> My whole reason for going into all of this is this was the only alternative that I have found in place of this method, problem is they no longer sell the device used in here, and the other methods do not transport well from pre-production machines to full production machines that are totally separate. (two identical servers being used for OBS video visual backbone machines)

Sorry, forgot to link the video in question. 
https://www.youtube.com/watch?v=GZEoss4XIgc

**Comment by Michael-E-B** [comment]
> So I could use xdotool to make up bindings that I could have recognised in the hotkey section of OBS, and then bind those to keys in Keebie, and then in theory OBS would see those bindings being used when Keebie invokes xdotool to the key combinations?

Not the way I would phrase it, but I think you've got the gist of it.

> Very rube goldberg in the way of going about it, but if the keybindings in xdotool are not complicated, say like just having a combo of keys and linking them to Keebie for their respective keys, that sounds like it would work in my case.

It's fairly simple really. It's just calling a command from Keebie and as I said before the Xdotool commands are straightforward.

> My whole reason for going into all of this is this was the only alternative that I have found in place of this method, problem is they no longer sell the device used in here, and the other methods do not transport well from pre-production machines to full production machines that are totally separate. (two identical servers being used for OBS video visual backbone machines)

I don't use OBS so this paragraph is a bit beyond me.

Is everything is working now so we can close this issue?

**Comment by azurusnova** [comment]
> Is everything is working now so we can close this issue?
I would like to think that it is, but I can see that I can set keys in Keebie as key combinations, but again I know we covered that. 

I think what would make this a little less painful is if there was a gui out there for xdotool to work on setting up these keybinds and other implementations. I found this out there, but not much else. I am hoping this will help in my case and for whomever else is trying to accomplish my method. 
https://github.com/sickcodes/xdotool-gui

> Keebie is meant to just run shell commands, scripts, and executables however if you want to have a key simulate other key(s) you can install xdotool and invoke it to simulate keypresses, mouse movements, and all sort of stuff. For instance I have KEY_LEFTMETA": "xdotool key Super_L+space" in my default.json layer to trigger a shortcut for my window manager.

And from what I am gathering here, is when I have the xdotool bindings setup how I want them, then I can go into the file its stored in (wherever that is going to be), and then paste them into the corresponding Keebie binding for that key, and then it should be able to emulate a combo key-press that should be seen by other programs properly. 

I am sorry about my ignorance about al of this, but you are helping immensely and despite my doofy way about all of this, I am learning, while stumbling over my own feet, I am learning. I also do hope this will help anyone else out there curious about this sort of setup.

**Comment by Michael-E-B** [comment]
I took a look at the video you linked, Keebie is an entirely different solution so this video is irrelevant.

**Comment by azurusnova** [comment]
> I took a look at the video you linked, Keebie is an entirely different solution so this video is irrelevant.

I understand the video was irrelevant, it was just where my fall into this rabbit hole started, and how it brought me here.

**Comment by Michael-E-B** [comment]
>And from what I am gathering here, is when I have the xdotool bindings setup how I want them, then I can go into the file its stored in (wherever that is going to be), and then paste them into the corresponding Keebie binding for that key, and then it should be able to emulate a combo key-press that should be seen by other programs properly.

Xdotool doesn't do any sort of binding (as far as I'm aware) it simply send "fake" keystrokes. What I would do is setup your shortcuts in OBS, then use `keebie.py` with `--add` to enter your xdotool command to simulate your shortcut and bind it to a key.

> I understand the video was irrelevant, it was just where my fall into this rabbit hole started, and how it brought me here.

I see.

**Comment by azurusnova** [comment]
https://linuxhint.com/xdotool_stimulate_mouse_clicks_and_keystrokes/

From what I can find in this link, it makes the setup of xdotool a little less intimidating, still going to be a heck of a shlaug to get everything set up how I want it to be. But I think after I get it set up how I want it, then I should be able to copy the Keebie folder with its respective files, and wherever the xdotool config is with its respective setup, and be able to move it over to a different machine with the keyboard I used, and it should (in theory) work. 

> What I would do is setup your shortcuts in OBS, then use keebie.py with --add to enter your xdotool command to simulate your shortcut and bind it to a key.

I am also going to test if that I setup a fake key command in xdotool and link it to a key in Keebie, then test to see if then pressing that key would make the corresponding fake keystroke show up in the options of OBS for the setup of the hotkeys. If that works, then I will be happy with that.

**Comment by Michael-E-B** [comment]
> ... then I should be able to copy the Keebie folder with its respective files, and wherever the xdotool config is with its respective setup, and be able to move it over to a different machine with the keyboard I used, and it should (in theory) work.

Xdotool doesn't have a config file, you just use it to send keystrokes (maybe you mean the Xdotool GUI).

> I am also going to test if that I setup a fake key command in xdotool and link it to a key in Keebie, then test to see if then pressing that key would make the corresponding fake keystroke show up in the options of OBS for the setup of the hotkeys. If that works, then I will be happy with that.

It definitely should, I have not yet found anything that misses Xdotool's fake keystrokes.

**Comment by azurusnova** [comment]
Let me see if I can do something simple as have the spacebar send an N keypress from the spacebar, and then see if the program will see that. I will report back on my findings.

**Comment by azurusnova** [comment]
> Let me see if I can do something simple as have the spacebar send an N keypress from the spacebar, and then see if the program will see that. I will report back on my findings.

I can report back that this method does work!

https://gitlab.com/cunidev/gestures/-/wikis/xdotool-list-of-key-codes
In the second column of this page with what I am assuming are the codes for each of these keys, how would I arrange it in a command when using 
"xdotool key entercodeshere"
when I want to use multiple key presses?

**Comment by Michael-E-B** [comment]
The first column are the keycodes you want to use, I think the second column is USB scancode, don't worry about them.

If by "multiple key presses" you mean pressing multiple keys at one you use `xdotool key keycode_1+keycode_2` with as many keycodes as you want, i.e. `xdotool key Control_L+c`.

**Comment by azurusnova** [comment]
I tried this and have confirmed that it does work. 
Thank you so much in showing me the way. 
I have backed up both the list of the keycode page and the "How to use xdotool" page as PDF's for my reference materials. 

Now if my friends didn't think I was a hacker before, this is going to blow their minds. XD
Again, thank you so much for your patience and your help. 
I really do hope this helps other noobs like myself. 
I learned a heck of a lot here.
